src/main/java/org/terasology/moduletestingenvironment/MainLoop.java

```diff
@@ -0,0 +1,211 @@
+        ChunkRegionFuture listener = ChunkRegionFuture.create(entityManager, relevanceSystem, centerBlock, desiredSize);
+        return listener.getFuture();
+    }
+
+    BlockRegionc chunkRegionToNewBlockRegion(BlockRegionc chunks) {
+        BlockRegion blocks = new BlockRegion(chunks);
+        return blocks.transform(new Matrix4f().scaling(new Vector3f(Chunks.CHUNK_SIZE)));
+    }
+
+    public <T> T runUntil(ListenableFuture<T> future) {
+        boolean timedOut = runUntil(future::isDone);
+        if (timedOut) {
+            // TODO: if runUntil returns timedOut but does not throw an exception, it
+            //     means it hit DEFAULT_GAME_TIME_TIMEOUT but not SAFETY_TIMEOUT, and
+            //     that's a weird interface due for a revision.
+            future.cancel(true);  // let it know we no longer expect results
+            throw new UncheckedTimeoutException("No result within default timeout.");
+        }
+        try {
+            return future.get(0, TimeUnit.SECONDS);
+        } catch (ExecutionException e) {
+            throw new UncheckedExecutionException(e);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interrupted while waiting for " + future, e);
+        } catch (TimeoutException e) {
+            throw new UncheckedTimeoutException(
+                    "Checked isDone before calling get, so this shouldn't happen.", e);
+        }
+    }
+
+    /**
+     * Runs tick() on the engine until f evaluates to true or DEFAULT_GAME_TIME_TIMEOUT milliseconds have passed in game time
+     *
+     * @return true if execution timed out
+     */
+    public boolean runUntil(Supplier<Boolean> f) {
+        return runWhile(() -> !f.get());
+    }
+
+    /**
+     * Runs tick() on the engine until f evaluates to true or gameTimeTimeoutMs has passed in game time
+     *
+     * @return true if execution timed out
+     */
+    public boolean runUntil(long gameTimeTimeoutMs, Supplier<Boolean> f) {
+        return runWhile(gameTimeTimeoutMs, () -> !f.get());
+    }
+
+    /**
+     * Runs tick() on the engine while f evaluates to true or until DEFAULT_GAME_TIME_TIMEOUT milliseconds have passed
+     *
+     * @return true if execution timed out
+     */
+    public boolean runWhile(Supplier<Boolean> f) {
+        return runWhile(ModuleTestingEnvironment.DEFAULT_GAME_TIME_TIMEOUT, f);
+    }
+
+    /**
+     * Runs tick() on the engine while f evaluates to true or until gameTimeTimeoutMs has passed in game time.
+     *
+     * @return true if execution timed out
+     */
+    public boolean runWhile(long gameTimeTimeoutMs, Supplier<Boolean> f) {
+        boolean timedOut = false;
+        Time hostTime = engines.getHostContext().get(Time.class);
+        long startRealTime = System.currentTimeMillis();
+        long startGameTime = hostTime.getGameTimeInMs();
+
+        while (f.get() && !timedOut) {
+            Thread.yield();
```

Comment:
Would it make sense to add some sort of throttled debug output here that could be used to show the time stats every, say, 10 seconds or so? To help identify when something goes wrong, like a weirdly long timeout, or maybe no progress being made (if there's something to ask for "status")

Reply:
We certainly have an ongoing need for _something_ to give us more visibility in to what's going on when a test stalls, but that's not in scope for this PR.